### PR TITLE
feat: LT-1304 - Team and Final Advice

### PR DESCRIPTION
**Pull request by cstobart**

### CODE REVIEW

File: cases/views/main.py [hunk 1]
```diff
@@ -260,14 +260,16 @@ def get(self, request, **kwargs):
...
-            'statuses': statuses
+            'statuses': reduced_statuses
         }
         return render(request, 'cases/manage.html', context)
 
     def post(self, request, **kwargs):
         case_id = str(kwargs['pk'])
         case, status_code = get_case(request, case_id)
 
+        print('\n\n\n', request.POST, '\n\n\n')
```
Comment (markm5A8Z9F43913A) [inline cases/views/main.py]: print

File: cases/views/advice.py [hunk 2]
```diff
@@ -1,159 +1,184 @@
...
+
+
+class Finalise(TemplateView):
+    def get(self, request, *args, **kwargs):
+        case, _ = get_case(request, str(kwargs['pk']))
+        advice, _ = get_final_case_advice(request, str(kwargs['pk']))
+        case_id = case['case']['id']
+
+        for item in advice['advice']:
+            print(item)
```
Comment (markm5A8Z9F43913A) [inline cases/views/advice.py]: print

File: cases/urls.py [hunk 3]
```diff
@@ -26,12 +26,30 @@
...
+    path('<uuid:pk>/user-advice-view/', advice.ViewUserAdvice.as_view(), name='user_advice_view'),
+    # ex: /cases/<uuid:pk>/team-advice-view/coalesce
+    path('<uuid:pk>/team-advice-view/coalesce', advice.CoalesceUserAdvice.as_view(), name='coalesce_user_advice'),
+    # ex: /cases/<uuid:pk>/team-advice-view/
+    path('<uuid:pk>/team-advice-view/', advice.ViewTeamAdvice.as_view(), name='team_advice_view'),
+    # ex: /cases/<uuid:pk>/final-advice-view/coalesce
+    path('<uuid:pk>/final-advice-view/coalesce', advice.CoalesceTeamAdvice.as_view(), name='coalesce_team_advice'),
+    # ex: /cases/<uuid:pk>/final-advice-view/
+    path('<uuid:pk>/final-advice-view/', advice.ViewFinalAdvice.as_view(), name='final_advice_view'),
     # ex: /cases/<uuid:pk>/advice-view/give-advice/
```
Comment (markm5A8Z9F43913A) [inline cases/urls.py]: Wrong example

File: cases/urls.py [hunk 4]
```diff
@@ -26,12 +26,30 @@
     path('<uuid:pk>/goods/<str:good_pk>/', goods.Good.as_view(), name='good'),
     # ex: /cases/<uuid:pk>/assign-flags/
     path('<uuid:pk>/assign-flags/', flags.AssignFlags.as_view(), name='assign_flags'),
-    # ex: /cases/<uuid:pk>/advice-view/
-    path('<uuid:pk>/advice-view/', advice.ViewAdvice.as_view(), name='advice_view'),
+    # ex: /cases/<uuid:pk>/user-advice-view/
+    path('<uuid:pk>/user-advice-view/', advice.ViewUserAdvice.as_view(), name='user_advice_view'),
+    # ex: /cases/<uuid:pk>/team-advice-view/coalesce
+    path('<uuid:pk>/team-advice-view/coalesce', advice.CoalesceUserAdvice.as_view(), name='coalesce_user_advice'),
```
Comment (markm5A8Z9F43913A) [inline cases/urls.py]: Is a trailing slash needed?

File: cases/urls.py [hunk 5]
```diff
@@ -26,12 +26,30 @@
...
+    # ex: /cases/<uuid:pk>/team-advice-view/
+    path('<uuid:pk>/team-advice-view/', advice.ViewTeamAdvice.as_view(), name='team_advice_view'),
+    # ex: /cases/<uuid:pk>/final-advice-view/coalesce
+    path('<uuid:pk>/final-advice-view/coalesce', advice.CoalesceTeamAdvice.as_view(), name='coalesce_team_advice'),
```
Comment (markm5A8Z9F43913A) [inline cases/urls.py]: Is a trailing slash needed?

File: cases/views_helpers.py [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+            'end_user': data.get('end_user'),
+            'ultimate_end_users': data.get('ultimate_end_users'),
+            'errors': response['errors'][0],
+            'data': data,
+            'level': user_team_final
+        }
+        return render(request, form, context)
+
+    # Add success message
+    messages.success(request, 'Your advice has been posted successfully')
```
Comment (markm5A8Z9F43913A) [inline cases/views_helpers.py]: Should this be in lite-content?

File: cases/forms/advice.py [hunk 7]
```diff
@@ -16,4 +16,4 @@ def advice_recommendation_form(case_id):
                 ],
                 default_button_name='Continue',
                 back_link=BackLink('Back to advice'),
-                post_url=reverse_lazy('cases:give_advice', kwargs={'pk': case_id}))
+                post_url=post_url)
```
Comment (andytaylorkainos) [inline cases/forms/advice.py]: 👍

File: cases/forms/finalise_case.py [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+        description='',
+        questions=[
+            DateInput(description='For example, 27 3 2007', title='When will the licence start?', prefix=''),
+            TextInput(name='duration', description='This must be a whole number of months, such as 12', title='How long will it last?'),
+        ],
+        back_link=BackLink(url=reverse_lazy('cases:final_advice_view', kwargs={'pk': case_id}), text='back to final advice'),
+    )
+
+
+def refuse_licence_form(case_id):
```
Comment (andytaylorkainos) [inline cases/forms/finalise_case.py]: maybe only pass in the post_url if its only needed, like above

File: cases/forms/finalise_case.py [hunk 9]
```diff
@@ -0,0 +1,22 @@
+from django.urls import reverse_lazy
+from lite_forms.components import Form, TextInput, BackLink, DateInput
+
+
+def approve_licence_form(case_id):
```
Comment (andytaylorkainos) [inline cases/forms/finalise_case.py]: maybe only pass in the post_url if its only needed, like above

File: cases/views/advice.py [hunk 10]
```diff
@@ -1,159 +1,184 @@
...
+            if item['type']['key'] == 'approve' or item['type']['key'] == 'proviso':
+                return form_page(request, approve_licence_form(case_id))
+
+        return form_page(request, refuse_licence_form(case_id))
+
+    def post(self, request, *args, **kwargs):
+        case, _ = get_case(request, str(kwargs['pk']))
+        application_id = case.get('case').get('application').get('id')
+        data = request.POST.copy()
```
Comment (andytaylorkainos) [inline cases/views/advice.py]: any reason why we need to specify a shallow copy?

File: cases/forms/finalise_case.py [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+    return Form(
+        title='Approve',
+        description='',
+        questions=[
+            DateInput(description='For example, 27 3 2007', title='When will the licence start?', prefix=''),
+            TextInput(name='duration', description='This must be a whole number of months, such as 12', title='How long will it last?'),
+        ],
+        back_link=BackLink(url=reverse_lazy('cases:final_advice_view', kwargs={'pk': case_id}), text='back to final advice'),
```
Comment (jfaracik) [inline cases/forms/finalise_case.py]: Capitalise the B in back

File: cases/forms/finalise_case.py [hunk 12]
```diff
@@ -0,0 +1,22 @@
+from django.urls import reverse_lazy
+from lite_forms.components import Form, TextInput, BackLink, DateInput
+
+
+def approve_licence_form(case_id):
+    return Form(
+        title='Approve',
+        description='',
```
Comment (jfaracik) [inline cases/forms/finalise_case.py]: Remove description

File: cases/forms/finalise_case.py [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+            TextInput(name='duration', description='This must be a whole number of months, such as 12', title='How long will it last?'),
+        ],
+        back_link=BackLink(url=reverse_lazy('cases:final_advice_view', kwargs={'pk': case_id}), text='back to final advice'),
+    )
+
+
+def refuse_licence_form(case_id):
+    return Form(
+        title='Refuse',
+        description='',
```
Comment (jfaracik) [inline cases/forms/finalise_case.py]: And here

File: cases/forms/finalise_case.py [hunk 14]
```diff
@@ -0,0 +1,22 @@
...
+        back_link=BackLink(url=reverse_lazy('cases:final_advice_view', kwargs={'pk': case_id}), text='back to final advice'),
```
Comment (jfaracik) [inline cases/forms/finalise_case.py]: B

File: ui_automation_tests/pages/application_page.py [hunk 15]
```diff
@@ -25,7 +25,7 @@ class ApplicationPage(BasePage):
     edit_case_flags = 'application-edit-case-flags'  # ID
     edit_goods_flags = 'button-edit-goods-flags'  # ID
     checkbox_input = ".govuk-checkboxes__input"
-    view_advice = "[href*='/advice-view/']"
+    view_advice = "a[href*='/user-advice-view/']"
```
Comment (jfaracik) [inline ui_automation_tests/pages/application_page.py]: Single quotes

File: cases/views_helpers.py [hunk 16]
```diff
@@ -0,0 +1,210 @@
...
+from core.services import get_denial_reasons, get_user_permissions
+from picklists.services import get_picklists
+from teams.services import get_teams
+from users.services import get_gov_user
+
+
+def get_case_advice(get_advice, request, case, user_team_final, team=None):
+    """
+    :param get_advice: This is a service method to get the advice from a particular level
+    :param request:
```
Comment (jfaracik) [inline cases/views_helpers.py]: Remove request comment if it isn't helpful

File: cases/views/main.py [hunk 17]
```diff
@@ -263,6 +268,7 @@ def post(self, request, **kwargs):
         if case['type']['key'] == 'application':
             application_id = case.get('application').get('id')
             _, _ = put_applications(request, application_id, request.POST)
```
Comment (jfaracik) [inline cases/views/main.py]: Remove _,_ =

File: ui_automation_tests/features/give_advice.feature [hunk 18]
```diff
@@ -72,3 +72,25 @@ Feature: I want to record my user advice and any comments and conditions relatin
...
+    And I choose to 'approve' the licence
+    And I import text from the 'advice' picklist
+    And I write 'We will get back to you in three weeks' in the note text field
+    And I click continue
+    And I go to the team advice
+    And I combine all advice
+    And I go to the final advice
+    And I combine all advice
+    And I finalise the licence
+    Then I reset the permissions
```
Comment (phlptt) [inline ui_automation_tests/features/give_advice.feature]: know it exists already but should have this as a fin hook like how we clear down the browser

File: ui_automation_tests/features/manage_cases.feature [hunk 19]
```diff
@@ -17,64 +17,6 @@ Feature: I want to record the final decision overall on an application case
...
-    And I select decision "2b"
-    And I type optional text "Reason denied due to bad information"
-    And I click continue
-    Then I see application "denied"
-    And I see denied reason
-    When I click record decision
-    And I click continue
-    And I select decision "2b"
-    And I click continue
-    Then I see application "denied"
```
Comment (phlptt) [inline ui_automation_tests/features/manage_cases.feature]: can we lose any of these from the step defs if they arent used anymore?

File: ui_automation_tests/features/give_advice.feature [hunk 20]
```diff
@@ -72,3 +72,25 @@ Feature: I want to record my user advice and any comments and conditions relatin
...
+
+  @LT_1115_grant
+  Scenario: Finalise a licence
+    Given I create application or application has been previously created
+    And I create a proviso picklist
+    And I create a standard advice picklist
+    And I sign in to SSO or am signed into SSO
+    When I give myself all permissions
+    And I go to application previously created
+    And I click on view advice
```
Comment (phlptt) [inline ui_automation_tests/features/give_advice.feature]: http://itsadeliverything.com/declarative-vs-imperative-gherkin-scenarios-for-cucumber
give this a read about declarative style. we are currently writing them in a bad style ie i click this i click that. makes it quicker to write and reuse but more difficult to maintain. going to go through everything and tidy it all up at some point but worth keeping in mind for future stories

**Pull request merged**